### PR TITLE
Fixes for verify-commits script

### PR DESCRIPTION
**Pull request by petertodd**
Tried to use it for another project and ran into some issues getting it to work on Debian 8

### CODE REVIEW

**Comment by MarcoFalke** [comment]
utACK   6d1016f

**Comment by laanwj** [comment]
utACK

File: contrib/verify-commits/verify-commits.sh [hunk 1]
```diff
@@ -1,9 +1,7 @@
-#!/bin/sh
+#!/bin/bash
 
-DIR=$(dirname "$0")
-
-echo "Please verify all commits in the following list are not evil:"
```
Comment (laanwj) [inline contrib/verify-commits/verify-commits.sh]: Did you intend to remove this message?

**Comment by laanwj** [comment]
Ping @TheBlueMatt

File: contrib/verify-commits/verify-commits.sh (hunk 1 above)
Comment (MarcoFalke) [inline contrib/verify-commits/verify-commits.sh]: Yes, it's the third commit.

**Comment by TheBlueMatt** [comment]
Which features are bash-specific (I'm pretty sure we went through this when it was merged, and there were none left)? We should just remove the bash-specific ones, not revert to forcing bash.

**Comment by TheBlueMatt** [comment]
Also, I do prefer to not remove the commits list message...its not there to prevent someone who has malicious access to the repo, its there to encourage people to audit the commits that were merged. If you prefer, just replace with it a notice to tell people to run git log on contrib/verify-commits manually.

File: contrib/verify-commits/gpg.sh [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
 INPUT=$(</dev/stdin)
```
Comment (petertodd) [inline contrib/verify-commits/gpg.sh]: @TheBlueMatt This doesn't seem to work on /bin/sh

**Comment by petertodd** [comment]
@TheBlueMatt The two constructs that don't seem to work with /bin/sh are:

https://github.com/petertodd/bitcoin/blob/2016-03-fix-verify-commits/contrib/verify-commits/verify-commits.sh#L28

https://github.com/petertodd/bitcoin/blob/2016-03-fix-verify-commits/contrib/verify-commits/gpg.sh#L2

**Comment by petertodd** [comment]
@TheBlueMatt Again, if you want to encourage people, do it in a README file or similar which doesn't give the wrong impression about what's the right way to do it.

Anyway, the necessity to audit the codebase *in general* should be something every security conscious developer should be aware of; calling out this program specifically may in itself give people the wrong impression.

**Comment by sipa** [comment]
utACK 6d1016f543837fe3c7b0a48573771f4b17e35ff5

**Comment by petertodd** [comment]
@sipa Sorry, just rebased and added one more commit.

**Comment by gmaxwell** [comment]
ACK e2764cb23826908bde171067fc5d0b0659b18952

**Comment by MarcoFalke** [comment]
utACK e2764cb

**Comment by TheBlueMatt** [comment]
NACK. Please fix the two bash-isms instead of requiring bash, and avoid adding more OS-specificisms with GNU-isms (which aren't even in some BSDs). Feel free to move the warnings to a README file of somesort, though.

**Comment by petertodd** [comment]
@TheBlueMatt I don't know enough about bash to know what needs to be fixed or how; if you'd like to suggest some changes please do. But if it's not going to get fixed I'd rather the failure be clear - "bash is needed and isn't installed" - than unclear.

**Comment by TheBlueMatt** [comment]
The three commits at https://github.com/TheBlueMatt/bitcoin/commits/2016-20-7713-fixes should replace your top three just fine.

**Comment by petertodd** [comment]
Incorporated @TheBlueMatt's /bin/sh fixes, so no longer depending on bash again.

**Comment by gmaxwell** [comment]
reACK 11164ec0b4c1790059220b09b5827e5618a46c76

**Comment by TheBlueMatt** [comment]
Please do leave the warning telling people to check git log in... making people see it every time until it's removed definitely has value. If not it should be added to a README somewhere.

**Comment by petertodd** [comment]
@TheBlueMatt Git has had vulnerabilities before ([example](https://www.mehmetince.net/one-git-command-may-cause-you-hacked-cve-2014-9390-exploitation-for-shell/)) where even just checking out source is sufficient to pwn your system; if you can run git log on the verify-commits directory it may be too late. If we're going to be giving advice on how to use verify-commits we should do it properly; for now removing that advice is very appropriate given that it's misleading at best.

**Comment by TheBlueMatt** [comment]
I find it rather unlikely that git log would have a vulnerability that cant also be triggered by some other checkout/pull/etc flow. I dont think telling people to run git log is bad advice here.
I agree the existing auto-run-git log isnt ideal, but I dont see why telling people to do so is bad, let alone why removing it entirely is better.

**Comment by laanwj** [comment]
Please add the warning back in. Seems otherwise we don't get agreement here to merge this. Its removal can be discussed separately.

**Comment by petertodd** [comment]
Added README explaining how to use verify-commits.sh safely; this should answer @TheBlueMatt's objection to removing the warning.

**Comment by sipa** [comment]
Tested ACK (checked out this branch in one worktree, master in another, verified master using the code from this branch).

**Comment by MarcoFalke** [comment]
Anything left to do here?

**Comment by petertodd** [comment]
@marcofalke IMO it's ready for merging.

**Comment by TheBlueMatt** [comment]
"This is an incomplete work in progress,"

It is? I'd say it works pretty well for its goals?

**Comment by TheBlueMatt** [comment]
Also you might want to note that you can run git log with paths to get a list of changes to that path in the README, just because it's useful for these scripts and not really an obvious feature that everyone is aware of.

**Comment by laanwj** [comment]
Can we please finish this?
This is pretty much a trivial change, but every time there comes up a new load of nits, which of those things is critical and can't be fixed later?

**Comment by MarcoFalke** [comment]
@laanwj I think we can improve the wording of the doc later.

**Comment by TheBlueMatt** [comment]
I mean I think the README is just wrong right now? We could have also taken the easy solution of splitting out the contentious change.

File: contrib/verify-commits/trusted-keys [hunk 3]
```diff
@@ -1,8 +1,5 @@
 71A3B16735405025D447E8F274810B012346C9A6
 1F4410F6A89268CE3197A84C57896D2FF8F0B657
```
Comment (MarcoFalke) [inline contrib/verify-commits/trusted-keys]: Nit: I think you can remove this line and clear allow-revsig-commits.

Comment (petertodd) [inline contrib/verify-commits/trusted-keys]: Sorry, which line specifically?

Good point re: allow-revsig

**Comment by petertodd** [comment]
@TheBlueMatt Sorry, but I'm kinda mystified why you think the README is wrong; I don't see it as controversial at all to describe this functionality as incomplete, given that we definitely need a convenient way for users to verify commits safely, prior to accidentally checking out repos and possibly running malicious code. I think it's good to point that out in the hope that others continue this starting point - and it's work that many other projects need as well (I personally use verify-commits in python-bitcoinlib, as well as some internal-use-only repos).

File: contrib/verify-commits/trusted-keys (hunk 3 above)
Comment (MarcoFalke) [inline contrib/verify-commits/trusted-keys]: `gpg --list-key --fingerprint 1F4410F6A89268CE3197A84C57896D2FF8F0B657` returns revoked key by @sipa. Once the commits signed by this key are buried under the trusted root, you could get rid of those lines in allow-revsig-commits.

**Comment by petertodd** [comment]
@MarcoFalke Fixed your nit re: sipa's old key.

**Comment by gmaxwell** [comment]
ACK 1e9aab0dbfcb844458b5f221a9af0141bba6280f

Readme seems fine to me.

**Comment by MarcoFalke** [comment]
utACK 1e9aab0.

I think the readme could be fixed later, if desired.

**Pull request merged**